pkg/pgmodel/cache.go

```diff
@@ -0,0 +1,61 @@
+// This file and its contents are licensed under the Apache License 2.0.
+// Please see the included NOTICE for copyright information and
+// LICENSE for a copy of the license.
+
+package pgmodel
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/timescale/timescale-prometheus/pkg/clockcache"
+)
+
+var (
+	// ErrEntryNotFound is returned when entry is not found.
+	ErrEntryNotFound = fmt.Errorf("entry not found")
+)
+
+type seriesCache struct {
+	series *clockcache.Cache
+}
+
+func (b *seriesCache) GetSeries(lset Labels) (SeriesID, error) {
+	result, ok := b.series.Get(lset.String())
+	if !ok {
+		return 0, ErrEntryNotFound
+	}
+	return result.(SeriesID), nil
+}
+
+func (b *seriesCache) SetSeries(lset Labels, id SeriesID) error {
+	b.series.Insert(lset.String(), id)
+	return nil
+}
+
+// MetricNameCache stores and retrieves metric table names in a in-memory cache.
+type MetricNameCache struct {
+	Metrics *clockcache.Cache
+}
+
+// Get fetches the table name for specified metric.
+func (m *MetricNameCache) Get(metric string) (string, error) {
+	result, ok := m.Metrics.Get(metric)
+	if !ok {
+		return "", ErrEntryNotFound
+	}
+	return result.(string), nil
+}
+
+// Set stores table name for specified metric.
+func (m *MetricNameCache) Set(metric string, tableName string) error {
+	// deep copy the strings so the original memory doesn't need to stick around
+	metricBuilder := strings.Builder{}
```

Comment:
is this really the best way to copy a string?

Reply:
yup, I think so 😞 
The other way I know is something like
```go
newbytes := make([]byte, len(str))
copy(newbytes, str)
return BytesToString(newbytes)

func BytesToString(bytes []byte) (s string) {
	if len(bytes) == 0 {
		return ""
	}

	str := (*reflect.StringHeader)(unsafe.Pointer(&s))
	str.Data = uintptr(unsafe.Pointer(&bytes[0]))
	str.Len = len(bytes)

	return s
}
```
which has the disadvantage of using `unsafe`, and is basically what `string.Builder` does internally anyway.